### ISSUE TITLE
No way to access fields during __init_subclass__

### ISSUE THREAD

**Issue by dmsfabiano** [pydantic/pydantic#5369]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

There is no way to iterate through the `fields` from a class method anymore without instantiation

For example, if I want to iterate through all the fields at definition time to assert all of them have examples, I can't because there is no `cls.__fields__` anymore or no way to access the names of the fields from `cls`

I am more than likely missing something

### Example Code

```Python
from pydantic import BaseModel, Field, root_validator

ParentModelEnforcer(BaseModel):

   def __init_subclass__(cls, *args, **kwargs):
       # Here, now we have nothing to pass to `checker`
       _checker(cls, {})

   @root_validator()
   def _checker(cls, _values):
       # values is empty when called from __init_subclass__
       pass

ChildClass(ParentModelEnforcer):
   test_field_good: str = Field(..., example="pretty_test")
   test_field_bad: str
```

### Python, Pydantic & OS Version

```Text
2.01a
```

**Comment by samuelcolvin** [comment]
There is no 2.04a release.

To answer the question, you need to use `.model_fields`.

**Comment by dmsfabiano** [comment]
Thank you for the super quick feedback! unless I am missing something `cls.model_fields` is also empty at definition class during `__init_subclass__`

**Comment by samuelcolvin** [comment]
Ye, we set it after the class is created, this is empty in `__init_subclass__`.

What are your trying to do?

**Comment by dmsfabiano** [comment]
Ok so basically, like I mentioned, I am trying to create a custom `MyBaseModel` that enforces whomever is inheriting from my custom class is complaint with certain arbitrary rules (validation during sub-classing)

For example, if I want to ensure they are defining an schema example in all the fields, I was able to do that even before instantiation (i.e. during `__init_subclass__`) using `cls.__fields__`, but now, I can't.

**Here is a pydantic v1 working example:** Note this is an oversimplified example but I think it should get the point across

```python
from pydantic import BaseModel, Field, root_validator

CustomModel(BaseModel):

   def __init_subclass__(cls, *args, **kwargs):
       _checker(cls, {})

   @root_validator()
   def _ensure_schema_has_examples(cls, _values):
       assert cls.Config.schema_extra["example"].keys() == cls.__fields__.keys()
       for field_name, field_value in cls.__fields__.items():
            # Field level validations and assertions
            pass

ChildClass(ParentModelEnforcer):
   test_field_good: str = Field(..., example="pretty_test")
   test_field_bad: str

  Config:
       schema_extra = {"example": {"test_field_good": "bad test"}}
```
In the example above, my `CustomModel` would catch this at definition time even before instantiating the object (at subclass time)

Now, in **v2** like I mentioned in the original post, the `cls` has obviously the the class attributes defined (i.e. `test_field_good`)
but since nowhere until the class is created `model_fields` (or any other class attribute) has a container of the fields, I can't perform these validations because I have no reference to be able to do `getattr`

**TLDR:** Essentially, there is no way to loop through the defined class attribute **fields** during `__init_subclass__` because `cls.__fields__` is gone and `cls.model_fields` is populated after the class is created/instantiated

**Comment by dmontagu** [comment]
It does seem reasonable to me to want something like `__init_subclass__` that can access the fields.

I see a few ways to achieve this:
* In `__init_subclass__`, just register the class for future handling, and then handle all the classes at some point prior to app launch. (This would work today, but is not a great experience for the user.)
* We could add a `__pydantic_init_subclass__` method that would be called immediately prior to returning `cls` in `ModelMetaclass.__new__`.

There is an interesting thread about this in the python mailing list: https://mail.python.org/archives/list/python-dev@python.org/thread/ZMRRNSFSLJZDGGZ66CFCYQBINU62CDNX/

Specifically, these snippets are directly related to the issue here:

> Looked at another way, `__init_subclass__` should be receiving a `cls` that is ready for use, i.e. done and fully 
constructed and complete. But if that class is being created by a custom metaclass, then the thing that is being given 
to `__init_subclass__` could easily be only partial.

> All in all I just worry about the backward compatibility here: one could
easily imagine a metaclass whose `__new__` relies on the effect of the
`__init_subclass__` called by type_new(), and moving the call would break
that. So it's probably 6 bugs fixed, half a dozen new bugs created. In such
a case, the status quo should win.

**Comment by dmontagu** [comment]
I'll also add that, if the call to `_model_construction.complete_model_class` at the end of `ModelMetaclass.__new__` was replaced by a call to some new classmethod, that would be an obvious place to perform "subclass validation" after calling the super()-version of the method.

(Admittedly, this is essentially exactly what a `__pydantic_init_subclass__` would be.)

**Comment by dmsfabiano** [comment]
Thank you so much @dmontagu , since it seems reasonable, could this be re-opened? or is there any other way, that I am missing to perform Model validation before instantiation?

Even if we add a super high level problem statement, like, "I want a parent pydantic Model that checks that all of its fields have examples", right now, I don't have a way to be able to parse through the fields from a model that would get sub-classed in a sharable class that could be re-used.

Unless I am missing something, I feel like not having much control in the __init_subclass__ does remove quite a bit of what we can do **before** instantiating the object

Example reason on why something like this may be needed before creating an object:
- I don't need to instantiate the model to check it's schema and validate what I would want
- I have no way to validate that the provided "examples" actually make sense! Say I have an API that uses Pydantic models as inputs and outputs, for my library, the user providing a valid example is crucial, and I would want to validate the example is correct. I shouldn't have to instantiate a random Model object to achieve this if the schema is already defined through class attributes

is there any class method that would help me with either of the above two?

**Comment by dmontagu** [comment]
@dmsfabiano the easiest way for you to do it without any changes would be to do something like:
```python
_types_to_validate = []

def validate_my_subtypes():
    for type_ in _types_to_validate:
        ... # do something to validate `type_`

class MyBaseModel(BaseModel):
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__()
        _types_to_validate.append(cls)



class MyModel(MyBaseModel):
    ...

... # define all your models

# once everything is ready:
validate_my_subtypes()
```
However, this is dependent on the user calling the `validate_my_subtypes` function. If you are creating a library and want to add additional validation, yeah, probably not currently possible without some funny hacks (or subclassing `ModelMetaclass`, which I would also strongly advise against).

I'll wait for @samuelcolvin to weigh before reopening the issue, but I think this is a very reasonable feature request and should be easy to support, so I (tentatively..) think we'll be able to put something in place for this.

**Comment by dmontagu** [comment]
@dmsfabiano I'm assuming the changes in #5386 would address your use case, would be good if you could confirm that before we merge it though. (Note it might still get some changes once Samuel has a chance to review it, but we discussed it and I think this will go through, just possibly with a change to the method name)

**Comment by dmsfabiano** [comment]
It does, thank you very much for the quick turnaround and support @dmontagu @samuelcolvin !

**Comment by dmontagu** [comment]
@dmsfabiano the "fix" for this from #5386 (`__pydantic_init_subclass__`) is included now in the just-released `pydantic==2.0a2`. Please let us know if you have any further issues with this.
